[RFC] Qt and qml access to “struct preferences” version 2.0

janiversen opened the pull request:
<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [ ] Functional change
- [x ] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->
This PR implements a new uniform set of classes to share struct preferences with Qt and QML as well as a uniform way to load/sync to local disk.

Why a new set of classes ? There are several reasons, that made me make this rather large investment:

- SettingsObjectWrapper contains a good class structure but has a number of hard to maintain features
A) The load/sync functions are separated and thus the storage names are duplicated, which leads to a potential problem.
B) the get/set/notify functions does not cover all variables in struct preferences, nor the needed variables in QML.
C) SettingsObjectWrapper is not available for QML (among others due to a private constructor) and QML set/get functions are duplicated in qmlmanager.
- QSettings and prefs-macros are used elsewhere in e.g. planner, making maintenance more “distributed”

The goal of the new classes are:
- present all of struct preferences to qt and QML.
- have a unified set/get/notify functions grouped logically (reused from SettingsObjectWrapper)
- have a single point to load/sync a variable ensuring data consistency
- do not change the local storage (backward compability)

Steps made:
1) A lot of the time went into researching the code and securing the consistency
2) Group/sort struct preferences to make it easier to find variables
3) make 1 class pr group of variables, with a load/sync function
4) ensure all classes are pure interface classes (this is needed because QML have a nasty tendency of creating new objects)

The current PR is complete seen from the data perspective, but needs more work before being merged. I publish the PR in order to get some feedback, and expect to make changes based on the feedback.

While we discuss this PR I am working on what is missing:
- remove the need for QSettings and prefs-macros outside settings
- convert the last 2 classes in subsurface-qt
- enhance the preferences test system
- add a dynamic qml page that (on demand) shows all variables in struct preferences (and possibly allow manipulation)
- remove the qml get/set functions outside settings (mainly qmlmanager)

### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
<!-- Ensure the test cases are updated if needed. -->

### Related issues:
<!-- Reference issues with #<issue-num>. -->
<!-- Write "Fixes #<issue-num" to notify Github that this PR fixes an issue. -->

### Additional information:
<!-- Include sample dive log or other relevant information to allow testing the change where feasible. -->

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->

### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->

### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. -->
@tcanabrava I understand you are the original author of SettingsObjectWrapper so I would especially like to hear your comments on this new set of classes, and ideas on how to make them even better
@dirkhh as written above, this PR is not to be merged in its current state, but hopefully once it is updated due to feedback and the missing parts.
@bstoeger I am not an expert on templates, but I have a feeling that the get/set/disk functions could be made simpler with templates ?

I look forward to a positive and prosperous discussion, which surely will lead to a better qml interface.

janiversen commented:
Just one short comment, apart from a few variables, it is really not renaming but writing new classes combinig information from several places in the current code (like e.g. the SettingsObjectWrapper does not provide qml access this is done in qmlmanager and qmlprefs).

I do see the point of your comments and will update the PR accordingly.

janiversen commented:
ooh forgot the change qmlPrefs to qPrefs was to signalize the classes are not only for qml, but lets keep the existing name.

janiversen commented on core/pref.h:
```diff
@@ -182,8 +223,6 @@ enum cloud_status {
 
 extern struct preferences prefs, default_prefs, git_prefs;
 
-#define PP_GRAPHS_ENABLED (prefs.pp_graphs.po2 || prefs.pp_graphs.pn2 || prefs.pp_graphs.phe)
```
this is “ups”. I have tested that every commit compiles as desktop, mobile and iOS. But this change should really have neen in the following commit, one of my git rebase did not do as I wanted. will correct it.

janiversen replied:
Done, it was a rebase mistake, ALL commits compile both as desktop and as mobile.

tcanabrava commented:
I wouldn't touch this unless it's for the better (in the sense "less code, more maintanability") - it took me quite a while to get everything right here (also in the sense "it's working"), before that the preferences where actually a huge distributed mess and dirk almost killed me when I proposed that rewrite.

Then I started creating code to automatically generate preferences code for me, take a look at https://github.com/tcanabrava/configuration-parser - if I would rework the preferences one more time I would use that (or a variant like KConfigXt).

Also, this can be exposed to Qml.

janiversen commented:
new version uploaded, where I think I have addressed all the questions/concerns. Renaming have been reduced to a bare minimum and the new classes not "just" copies, but adds functionality hence updated commit messages.

I also changed the Q_PROPERTY formatting, although nobody commented it in #1378.

I hope "I don't think I can see the merit of the PR" from @dirkhh was not really correct.

janiversen commented on profile-widget/profilewidget2.cpp:
```diff
@@ -44,6 +44,8 @@
 #endif
 #include <QtWidgets>
 
+#define PP_GRAPHS_ENABLED (prefs.pp_graphs.po2 || prefs.pp_graphs.pn2 || prefs.pp_graphs.phe)
```
done

janiversen commented on core/subsurface-qt/SettingsObjectWrapper.h:
```diff
@@ -14,39 +14,6 @@
-	Q_PROPERTY(QString product READ dc_product WRITE setProduct NOTIFY productChanged)
-	Q_PROPERTY(QString device READ dc_device WRITE setDevice NOTIFY deviceChanged)
-	Q_PROPERTY(QString device_name READ dc_device_name WRITE setDeviceName NOTIFY deviceNameChanged)
-	Q_PROPERTY(int download_mode READ downloadMode WRITE setDownloadMode NOTIFY downloadModeChanged)
```
Well the other form was ok in #1378, not a single comment there. apart from that, I hate long lines. but I have changed them all, since this is not an important issue for me.

janiversen commented on core/CMakeLists.txt:
```diff
@@ -29,8 +29,9 @@ if(BLESUPPORT)
 endif()
 
-# compile the core library, in C.
+# compile the core library part in C, part in C++
```
Well it seems rather random as if people have added files here and there. I am sorry that I could not see the logic.

janiversen commented on core/settings/qPref_private.h:
```diff
@@ -0,0 +1,75 @@
+{		\
+	free((void *)prefs.name);	\
+	prefs.name = copy_qstring(string);	\
+}
```
prefs-macros was used before.

janiversen commented:
Closing this now, as I am presenting it in pieces.
